server/src/main/java/org/elasticsearch/search/profile/AbstractProfileBreakdown.java

```diff
@@ -25,50 +28,58 @@ public abstract class AbstractProfileBreakdown<T extends Enum<T>> {
     /**
      * The accumulated timings for this query node
      */
-    private final Timer[] timings;
-    private final T[] timingTypes;
+    private final Map<T, List<Timer>> timings;
 
     /** Sole constructor. */
     public AbstractProfileBreakdown(Class<T> clazz) {
-        this.timingTypes = clazz.getEnumConstants();
-        timings = new Timer[timingTypes.length];
-        for (int i = 0; i < timings.length; ++i) {
-            timings[i] = new Timer();
+        T[] enumConstants = clazz.getEnumConstants();
+        timings = Maps.newMapWithExpectedSize(enumConstants.length);
```

Comment:
why not use a concurrent hashmap instead of synchronizing manually?

Reply:
Will do

Reply:
Would concurrent hashmap actually work? That only synchronizes hashmap operations and we don't ever add/remove keys?

It seems the synchronization only occurs for the contained lists of timers.